### PR TITLE
NEXT-37141 - Add button for generation of one or more variants without deleting existing ones

### PR DESCRIPTION
**Pull request by amenk**
### 1. Why is this change necessary?

There is a currently no easy way to add single variant combinations. When trying to do so, using the "Generate Variants" function will also delete variants, you still might need.

### 2. What does this change do, exactly?

Adds a new mode "onlyAdd" to the modal which will not remove the variants.

Demo:

variants-poc

### 3. Describe each step to reproduce the issue or behaviour.

1. Go to a variant product
2. Click add variants
3. Select some combinations
4. Click next
5. Variants are only added, not deleted

### 5. Checklist

- [ ] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### Todos (help wanted)

- [x] Clean the modal preselection when variants are added
- [ ] Jest Tests
- [ ] Fix Button's CSS

### <samp>🤖 Generated by Copilot at d7ad7d9</samp>

This pull request adds a new feature to the product module, which allows adding variants without deleting existing ones. It adds a new `isAddOnly` option to the `sw-products-variants-generator` helper and the `sw-product-modal-variant-generation` component, and modifies the product detail page and the variant generation modal to use it. It also updates the labels and tooltips of the buttons for generating and adding variants, and the corresponding translations.

### <samp>🤖 Generated by Copilot at d7ad7d9</samp>

*  Add a new prop `isAddOnly` to the `sw-product-modal-variant-generation` component to control whether the modal should only add new variants or also synchronize and delete existing ones ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-ea420e5967d93facf6954d8f20ca93259a2188bd84a559675fe5d07051128ebcR44-R49), [link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-92dcd7efc2b93dbf1209d9ef26ddfeaefb2d11f3a80dfa5777e4fa36afc1af78R156), [link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-92dcd7efc2b93dbf1209d9ef26ddfeaefb2d11f3a80dfa5777e4fa36afc1af78L160-R177))
*  Pass the `isAddOnly` prop to the `sw-products-variants-generator` helper, which generates and filters the variant permutations based on the product configuration and the existing variants on the server ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-ea420e5967d93facf6954d8f20ca93259a2188bd84a559675fe5d07051128ebcR311), [link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-c0b9359b64314f80afca119ec51bf6ddbed2c94d585aa85af046d5dfecb16277R54), [link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-c0b9359b64314f80afca119ec51bf6ddbed2c94d585aa85af046d5dfecb16277L85-R86), [link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-c0b9359b64314f80afca119ec51bf6ddbed2c94d585aa85af046d5dfecb16277R97))
*  Modify the `filterVariations` method of the `sw-products-variants-generator` helper to skip the creation of the delete queue if the `isAddOnly` flag is true ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-c0b9359b64314f80afca119ec51bf6ddbed2c94d585aa85af046d5dfecb16277L123-R135))
*  Add a new button for adding variants to the `sw-product-variants-overview` component, which emits the `add-open` event to open the modal with the `isAddOnly` prop set to true ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-ae3259fdedfe62137ddcc40ced60bd72f2af106af8aa8b8cc4d44b19db27fb5dL112-R130))
*  Change the label of the existing button for generating variants to "Synchronize / Generate" in the `sw-product-variants-overview` component, which emits the `generator-open` event to open the modal with the `isAddOnly` prop set to false ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-ae3259fdedfe62137ddcc40ced60bd72f2af106af8aa8b8cc4d44b19db27fb5dL112-R130))
*  Add a tooltip for the add button in the `sw-product-variants-overview` component, which shows a warning message if the user does not have the `product.creator` privilege ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-ae3259fdedfe62137ddcc40ced60bd72f2af106af8aa8b8cc4d44b19db27fb5dL112-R130))
*  Add a listener for the `add-open` event to the `sw-product-detail-variants` component, which calls the `openModal` method with the `variantAdd` argument to show the modal for adding variants ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-92dcd7efc2b93dbf1209d9ef26ddfeaefb2d11f3a80dfa5777e4fa36afc1af78R140))
*  Update the `sw-product` snippet file for the `en-GB` locale, which contains the translations for the product module. Change the translation for the `generateVariants` key to "Synchronize / Generate" and add a new translation for the `addVariants` key as "Add" ([link](https://github.com/shopware/shopware/pull/3416/files?diff=unified&w=0#diff-38ba9297d9626e37eada8d80e7245c2320e74a099152d2210099a45d1b5ef283L385-R386))

### CODE REVIEW

**Comment by OliverSkroblin** [comment]
Hello @amenk,

That should be already the case, right? If I go into admin and just add more options, it just generates new variants.

Scenario:
- Generate variants with “color: red, green”.
- 2x variants were generated
- Now change the configuration to “Color: Red, Green, White, Black”.
- Black + White is generated.

Only when I add another group (e.g. Size) are the existing variants deleted and completely regenerated.

**Comment by amenk** [comment]
@OliverSkroblin 
Thanks for the feedback.
Actually the MR is missing one important functionality - when the "Add" button is used, the property selection tree should start with no preselection.

This is for usecases, where not all variants exist in all combinations. In our case it's seminar date, but with the T-Shirt case let's say you have a Cat-Print and a Dog-Print and the prints are not available in all colors.

So with the current implementation in the core you can only add

* Red, Green
* Cat, Dog

which will create 4 T-Shirts.

We want to be able to add first Red + Cat and then Green - Dog.

There is already an exclude configuration in the variant generate, but this is very cumbersome if you for example also have sizes S, M, L, XL and want to add configurations one-by-one.

Does it make this more clear? Let me know, if you have further questions.

**Comment by OliverSkroblin** [comment]
Okay, this makes sense, so you don't want to trigger the whole combination logic, you only want to add the combinations that are selected when the windows pop up.

In this case, the changes make sense to me ;)

**Comment by amenk** [comment]
@OliverSkroblin The problem we are currently facing in the implementation to clear the whole selection when using the add-button without breaking the normal sync-style-generation because the combinations data structure is tied tight to the dialogue.

Also I am wondering if both modes should coexist or - once the merchant chose to add single variants, the old option should be disabled, because when using the sync-button, also the missing variants would be generated anyways.

The same problem already happens if the merchant deletes single variant manually - the generator would always try to re-add them.

Do you have some advice for us?

**Comment by amenk** [comment]
I am thinking to add a button in the modal "switch to add-only mode" which will reset the selections and later only add. Because the selections are reset, the sync mode does not make much sense anymore. But I am not sure if merchants would want to use both modes anyways.

**Comment by amenk** [comment]
I have ported more code from the plugin solution we had in 6.4 
The selections are now properly cleared (for "add" and restored)

The merge request is basically feature-complete and ready for a first review.

**Comment by amenk** [comment]
Changelog added, snippets cleaned up and avoided renaming the old snippet. some linting errors seem unrelated to the PR .. let's see ..

**Comment by amenk** [comment]
How do we proceed on this? I already ported it to 6.6 in the last force push

File: src/Administration/Resources/app/administration/src/module/sw-product/helper/sw-products-variants-generator.js [hunk 1]
```diff
@@ -124,8 +126,21 @@ export default class VariantsGenerator extends EventEmitter {
...
-            // Copy the hashed list with the sorted variations on the server.
-            let deleteQueue = deepCopyObject(hashed);
+            let deleteQueue = [];
+            if (!isAddOnly) {
+                // Copy the hashed list with the sorted variations on the server.
+                deleteQueue = deepCopyObject(hashed);
+            } else {
+                // Reassign product numbers
+                let increment = 1;
+                createQueue.forEach((variant) => {
```
Comment (nguyenky) [inline src/Administration/Resources/app/administration/src/module/sw-product/helper/sw-products-variants-generator.js:136]: I tested this function, and it works fine. However, while adding Jest tests, I discovered that this specific code is never applied because variable `createQueue` is always empty  (as defined above `createQueue = []` ).

Could you please explain the purpose of this code? 
Will there be any negative impact if I delete it?
So far, I have tried deleting it, and everything seems to be working fine.

Comment (nguyenky) [reply src/Administration/Resources/app/administration/src/module/sw-product/helper/sw-products-variants-generator.js:136]: cc @amenk

Comment (amenk) [reply src/Administration/Resources/app/administration/src/module/sw-product/helper/sw-products-variants-generator.js:136]: Thank you very much for the review and catching this issue @nguyenky 

Some background: 

We ported the code originally from an internal module, and introduced this logic for the following reasons:

        /* The original function assigns the product number based on the fact that some variants might be deleted,
         * but we do not actually delete any so we have to take all variation on the server into account to avoid
         * duplicate product numbers.
         */

But as you say, it's currently not called. If all tests go well, let's remove it for good. I also did not see this problem again in testing - but please keep this case in mind when adding test cases - maybe some edge case is not covered yet in the ported code.

**Comment by nguyenky** [comment]
Thank you for your contribution @amenk ,

I wanted to inform you that the PR has been merged. I have made a few updates, so there are slight differences in the UI and behavior, but the core feature remains the same.

You can see my demo video and compare with your demo video above:


https://github.com/user-attachments/assets/4ef02eec-9e9b-49c3-8e57-46f8a640c4d5



This feature should be available in the next major release.

thanks, again

**Comment by amenk** [comment]
@nguyenky thank you very much - the new UI is awesome.

**Comment by amenk** [comment]
@nguyenky Actually I just noticed a problem, when reviewing the video once again. 

Video 0:20 - antiquewhite cannot be deselected

In our original implementation, all checkboxes were left empty when the "add only" mode is chosen. This is on purpose.

Now, anitquewhite cannot be deselected.

Additional testcase should be:

The user wants to add a single variant combination like

* color: aliceblue
* length: 1m

anymore, as we cannot deselect antiquewhite it always would add aliceblue/1m + anitquewhite/1m and even more combinations with the "content" attribute choices.

What do you think? Can all checkboxes be cleared, when the "add only" mode is chosen?

**Comment by amenk** [comment]
@nguyenky ping - what do you think?

**Comment by amenk** [comment]
Workarkound: 
* Uncheck all the other attributes
* Then check "add only"
* Now check the attributes you want to add

But this UX should be improved.

Eventually by clearing all attribute selections when "add only" is checked.

**Comment by amenk** [comment]
@nguyenky Hey, do you have any update on this from the UX perspective? Shall I create a new issue?

**Comment by nguyenky** [comment]
@amenk It's currently under review. There are other ideas from our designer, so it won't be resolved quickly. I'll let you know if there are any changes

**Comment by nguyenky** [comment]
Hi @amenk ,

An adjustment merge request about UX  has been merged. This feature should be available in the next major release.

Thanks for your contribution @amenk

**Comment by amenk** [comment]
Hey @nguyenky

Works flawlessly - thank you very much!

[Screencast add Single Variants, Selection is cleared on add](https://github.com/user-attachments/assets/eb6eb0d6-d06f-46f7-80a1-b6ac6a916b36)
